rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp

```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
     const std::string & node_name,
     const std::string & namespace_) const = 0;
 
+  /// Return a map of existing service names and types with a specific node.
+  /**
+   * This function only considers clients - not service servers.
+   * The returned names are the actual names used and do not have remap rules applied.
+   *
+   * \param[in] node_name name of the node
+   * \param[in] namespace_ namespace of the node
+   */
+  RCLCPP_PUBLIC
+  virtual
+  std::map<std::string, std::vector<std::string>>
+  get_client_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_) const = 0;
+
+  /// Return a map of existing topic names to list of topic types for a specific node.
+  /**
+   * This function only considers publishers - not subscribers.
+   * The returned names are the actual names used and do not have remap rules applied.
```

Comment:
right, that is confusing for me too.

probably `rcl/graph` and `rcl_action/graph` comments are also confusing?

https://github.com/ros2/rcl/blob/b7784eb5d2fbff33053b69f6b9f9d3b703f9f2a7/rcl/include/rcl/graph.h#L76-L78

https://github.com/ros2/rcl/blob/b7784eb5d2fbff33053b69f6b9f9d3b703f9f2a7/rcl_action/include/rcl_action/graph.h#L38-L40

Reply:
> probably rcl/graph and rcl_action/graph comments are also confusing?

What that means is that you're getting that already remapped name, but if you pass that to `create_publisher()` the result might be wrong because of remapping rules being applied again.

The ` The returned names are not automatically remapped by this function.` makes the comment a bit confusing, so if you want to rephrase that that would be ok.
But the second one is the really important one.

P.S.: We should actually have a way to create a topic with **exactly** a name, which isn't currently possible.

Reply:
> But the second one is the really important one.

okay, i will add comments about this `do not have remap rules applied` part.